Sporadic crash during C-x o

I've seen this crash about a dozen times:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x2 pc=0x5722a9]

goroutine 1 [running]:
main.switchWindow()
        /home/username/go/src/github.com/ixtenu/gomacs/window.go:101 +0x29
main.LoadDefaultCommands.func20(0x3)
        /home/username/go/src/github.com/ixtenu/gomacs/commands.go:284 +0x17
main.(*CommandFunc).Run(0xc000738f20, 0xc0004b03e1)
        /home/username/go/src/github.com/ixtenu/gomacs/commands.go:193 +0x36
main.RunCommandForKey({0xc0004b03e1, 0x3}, 0xc000054f10)
        /home/username/go/src/github.com/ixtenu/gomacs/main.go:587 +0x2c5
main.main()
        /home/username/go/src/github.com/ixtenu/gomacs/main.go:758 +0x79d
```

Unfortunately, it's tricky to reproduce.  It always happens during `C-x o` (or shortly afterward) when `gomacs` has multiple windows viewing different files.  It seems to usually occur when `C-x o` immediately follows another windowing operation.  Sometimes, it doesn't crash right away, but the cursor disappears and the next `C-x o` crashes.

Any idea what might be going on here?  I assume the call to `getFocusWindow()`  in `switchWindow()` is returning `nil`, leading to a nil pointer dereference at window.go:101.  Maybe a race condition?  `getFocusWindow()` being called when `winTree.focused` is being moved from one `winTree` to another?

The above crash was with commit 65fb9ab1 (tip of the `master` branch).  I also saw it on the `zygomys` branch (commit 711b1216fd).  The first time I saw it was on February 6th (commit 47f7e1c7ce), before any of the recent changes.  The call stack is substantially the same each time, except for the pointer addresses.

--- a/commands.go
+++ b/commands.go
@@ -96,7 +96,7 @@ func (c *CommandList) GetCommand(key string) (*CommandFunc, error) {
 		return nil, errors.New("Bad command: " + Global.Input)
 	}
 	if child.Parent {
-		nextkey, _ := editorGetKey()
+		nextkey := editorGetKey()
 		s, e := child.GetCommand(nextkey)
 		return s, e
 	} else {
@@ -123,7 +123,7 @@ func DescribeKeyBriefly() {
 	defer editorSetPrompt("")
 	Global.Input = ""
 	editorRefreshScreen()
-	key, _ := editorGetKey()
+	key := editorGetKey()
 	if Global.MajorBindings[Global.CurrentB.MajorMode] != nil {
 		com, comerr := Global.MajorBindings[Global.CurrentB.MajorMode].GetCommand(key)
 		if com != nil && comerr == nil {

--- a/input.go
+++ b/input.go
@@ -17,15 +17,15 @@ func InitTerm() {
 	termbox.SetInputMode(termbox.InputAlt | termbox.InputMouse)
 }
 
-func editorGetKey() (string, bool) {
+func editorGetKey() string {
 	for {
 		ev := termbox.PollEvent()
 		if ev.Type == termbox.EventResize {
 			editorRefreshScreen()
 		} else if ev.Type == termbox.EventKey {
-			return ParseTermboxEvent(ev), true
+			return ParseTermboxEvent(ev)
 		} else if ev.Type == termbox.EventMouse {
-			return ParseMouseEvent(ev), true
+			return ParseMouseEvent(ev)
 		}
 	}
 }
@@ -299,7 +299,7 @@ func tabCompletedEditorPrompt(prompt string, getCandidates func(string) []string
 				for undecided {
 					Global.Input = candidates[choice]
 					editorRefreshScreen()
-					key, _ := editorGetKey()
+					key := editorGetKey()
 					if key == "TAB" || key == "C-i" || key == "RIGHT" || key == "C-f" {
 						choice++
 						if choice == len(candidates) {

--- a/input.go
+++ b/input.go
@@ -31,7 +31,7 @@ func editorGetKey() (string, bool) {
 		go func() {
 			select {
 			case _ = <-timeout:
-				doReHl = true
+				editorRefreshScreen()
 			case _ = <-done:
 				// Don't refresh the screen.
 			}

--- a/input.go
+++ b/input.go
@@ -3,7 +3,6 @@ package main
 import (
 	"fmt"
 	"strings"
-	"time"
 	"unicode/utf8"
 
 	termutil "github.com/japanoise/termbox-util"
@@ -20,30 +19,13 @@ func InitTerm() {
 
 func editorGetKey() (string, bool) {
 	for {
-		// More hacking; if we've been waiting for some time, refresh the screen.
-		doReHl := false
-		timeout := make(chan bool, 1)
-		done := make(chan bool, 1)
-		go func() {
-			time.Sleep(time.Duration(TIMEOUT))
-			timeout <- true
-		}()
-		go func() {
-			select {
-			case _ = <-timeout:
-				editorRefreshScreen()
-			case _ = <-done:
-				// Don't refresh the screen.
-			}
-		}()
 		ev := termbox.PollEvent()
-		done <- true
 		if ev.Type == termbox.EventResize {
 			editorRefreshScreen()
 		} else if ev.Type == termbox.EventKey {
-			return ParseTermboxEvent(ev), doReHl
+			return ParseTermboxEvent(ev), true
 		} else if ev.Type == termbox.EventMouse {
-			return ParseMouseEvent(ev), doReHl
+			return ParseMouseEvent(ev), true
 		}
 	}
 }

--- a/lisp.go
+++ b/lisp.go
@@ -10,11 +10,8 @@ import (
 )
 
 func lispGetKey(env *glisp.Glisp, name string, args []glisp.Sexp) (glisp.Sexp, error) {
-	key, drhl := editorGetKey()
-	if drhl {
-		editorRefreshScreen()
-		Global.CurrentB.updateHighlighting()
-	}
+	key, _ := editorGetKey()
+	editorRefreshScreen()
 	return glisp.SexpStr(key), nil
 }
 

--- a/lisp.go
+++ b/lisp.go
@@ -10,7 +10,7 @@ import (
 )
 
 func lispGetKey(env *glisp.Glisp, name string, args []glisp.Sexp) (glisp.Sexp, error) {
-	key, _ := editorGetKey()
+	key := editorGetKey()
 	editorRefreshScreen()
 	return glisp.SexpStr(key), nil
 }

--- a/macro.go
+++ b/macro.go
@@ -32,18 +32,15 @@ func micromode(repeatkey string, msg string, env *glisp.Glisp, f func(*glisp.Gli
 	f(env)
 	Global.Input = msg
 	editorRefreshScreen()
-	key, drhl := editorGetKey()
+	key, _ := editorGetKey()
 	for key == repeatkey {
 		f(env)
 		editorRefreshScreen()
-		key, drhl = editorGetKey()
+		key, _ = editorGetKey()
 	}
 	Global.SetUniversal = false
 	RunCommandForKey(key, env)
-	if drhl {
-		editorRefreshScreen()
-		Global.CurrentB.updateHighlighting()
-	}
+	editorRefreshScreen()
 }
 
 func doRunMacro(env *glisp.Glisp) {

--- a/macro.go
+++ b/macro.go
@@ -32,11 +32,11 @@ func micromode(repeatkey string, msg string, env *glisp.Glisp, f func(*glisp.Gli
 	f(env)
 	Global.Input = msg
 	editorRefreshScreen()
-	key, _ := editorGetKey()
+	key := editorGetKey()
 	for key == repeatkey {
 		f(env)
 		editorRefreshScreen()
-		key, _ = editorGetKey()
+		key = editorGetKey()
 	}
 	Global.SetUniversal = false
 	RunCommandForKey(key, env)

--- a/main.go
+++ b/main.go
@@ -54,8 +54,6 @@ type EditorBuffer struct {
 	Highlighter  *highlight.Highlighter
 	MajorMode    string
 	prefcx       int
-	rehlfrom     int
-	needshl      bool
 	regionActive bool
 	region       *Region
 }
@@ -156,23 +154,9 @@ func rowUpdateRender(row *EditorRow) {
 
 func editorReHighlightRow(row *EditorRow, buf *EditorBuffer) {
 	if buf.Highlighter != nil {
-		if buf.rehlfrom >= 0 && buf.rehlfrom != row.idx {
-			if row.idx < buf.rehlfrom {
-				buf.rehlfrom = row.idx
-			}
-			buf.needshl = true
-		} else {
-			buf.rehlfrom = row.idx
-		}
-	}
-}
-
-func (buf *EditorBuffer) updateHighlighting() {
-	if buf.Highlighter != nil && buf.rehlfrom >= 0 && buf.NumRows != 0 {
-		row := buf.Rows[buf.rehlfrom]
 		curstate := buf.State(row.idx)
 		buf.Highlighter.ReHighlightStates(buf, row.idx)
-		if curstate != buf.State(row.idx) || buf.needshl {
+		if curstate != buf.State(row.idx) {
 			// If the EOL state changed, the buffer needs rehighlighting
 			// as this was probably multiline comment or string.
 			buf.Highlighter.HighlightMatches(buf, row.idx, buf.NumRows)
@@ -180,8 +164,6 @@ func (buf *EditorBuffer) updateHighlighting() {
 			// Probably only this line changed.
 			buf.Highlighter.ReHighlightLine(buf, row.idx)
 		}
-		buf.rehlfrom = -1
-		buf.needshl = false
 	}
 }
 

--- a/main.go
+++ b/main.go
@@ -22,8 +22,6 @@ import (
 	"github.com/zyedidia/highlight"
 )
 
-const TIMEOUT int64 = 10000000
-
 type EditorRow struct {
 	idx        int
 	Size       int
@@ -748,7 +746,7 @@ func main() {
 			RunCommandForKey(key, env)
 			// A bit hacky, but this fixes some of our speed issues when pasting.
 			// Don't do the optimisation if this key and the last were the same!
-			if t.UnixNano()-lt.UnixNano() > TIMEOUT || lastkey == key {
+			if t.UnixNano()-lt.UnixNano() > 10000000 || lastkey == key {
 				editorRefreshScreen()
 			}
 			if drhl {

--- a/main.go
+++ b/main.go
@@ -582,7 +582,7 @@ func AddErrorMessage(msg string) {
 func SetUniversalArgument(env *glisp.Glisp) {
 	arg := ""
 	for {
-		key, drhl := editorGetKey()
+		key, _ := editorGetKey()
 		if (arg == "" && key == "-") || ('0' <= key[0] && key[0] <= '9') {
 			arg += key
 			Global.Input += key
@@ -591,7 +591,7 @@ func SetUniversalArgument(env *glisp.Glisp) {
 			if key == "C-u" {
 				Global.Input += " " + key
 				editorRefreshScreen()
-				key, drhl = editorGetKey()
+				key, _ = editorGetKey()
 			}
 			argi := 0
 			if arg != "" {
@@ -605,10 +605,7 @@ func SetUniversalArgument(env *glisp.Glisp) {
 			Global.Universal = argi
 			Global.SetUniversal = true
 			RunCommandForKey(key, env)
-			if drhl {
-				editorRefreshScreen()
-				Global.CurrentB.updateHighlighting()
-			}
+			editorRefreshScreen()
 			Global.SetUniversal = false
 			return
 		}
@@ -739,7 +736,7 @@ func main() {
 		if Global.quit {
 			return
 		} else {
-			key := editorGetKey()
+			key, _ := editorGetKey()
 			RunCommandForKey(key, env)
 		}
 	}

--- a/main.go
+++ b/main.go
@@ -582,7 +582,7 @@ func AddErrorMessage(msg string) {
 func SetUniversalArgument(env *glisp.Glisp) {
 	arg := ""
 	for {
-		key, _ := editorGetKey()
+		key := editorGetKey()
 		if (arg == "" && key == "-") || ('0' <= key[0] && key[0] <= '9') {
 			arg += key
 			Global.Input += key
@@ -591,7 +591,7 @@ func SetUniversalArgument(env *glisp.Glisp) {
 			if key == "C-u" {
 				Global.Input += " " + key
 				editorRefreshScreen()
-				key, _ = editorGetKey()
+				key = editorGetKey()
 			}
 			argi := 0
 			if arg != "" {
@@ -736,7 +736,7 @@ func main() {
 		if Global.quit {
 			return
 		} else {
-			key, _ := editorGetKey()
+			key := editorGetKey()
 			RunCommandForKey(key, env)
 		}
 	}

--- a/main.go
+++ b/main.go
@@ -12,7 +12,6 @@ import (
 	"runtime/pprof"
 	"strconv"
 	"strings"
-	"time"
 	"unicode/utf8"
 
 	termutil "github.com/japanoise/termbox-util"
@@ -734,26 +733,14 @@ func main() {
 
 	InitTerm()
 	defer termbox.Close()
-	editorRefreshScreen()
-	lastkey := "<none>"
-	lt := time.Now()
+
 	for {
+		editorRefreshScreen()
 		if Global.quit {
 			return
 		} else {
-			key, drhl := editorGetKey()
-			t := time.Now()
+			key := editorGetKey()
 			RunCommandForKey(key, env)
-			// A bit hacky, but this fixes some of our speed issues when pasting.
-			// Don't do the optimisation if this key and the last were the same!
-			if t.UnixNano()-lt.UnixNano() > 10000000 || lastkey == key {
-				editorRefreshScreen()
-			}
-			if drhl {
-				Global.CurrentB.updateHighlighting()
-				editorRefreshScreen()
-			}
-			lt = time.Now()
 		}
 	}
 }

--- a/registers.go
+++ b/registers.go
@@ -121,7 +121,7 @@ func (r *RegisterList) jumpToPositionRegister(regname string) {
 func InteractiveGetRegister(prompt string) (*Register, string) {
 	Global.Input = prompt
 	editorRefreshScreen()
-	reg, _ := editorGetKey()
+	reg := editorGetKey()
 	Global.Input += reg
 	return Global.Registers.Registers[reg], reg
 }

